### ISSUE TITLE
Run downloads in parallel/asynchronously

### ISSUE THREAD

**Issue by camelid** [rust-lang/cargo-bisect-rustc#120]
I think this would make bisecting faster.

The only downside I see is that it would be more difficult to have progress bars for each of the components. Perhaps we could have one progress bar that shows the *total* number of bytes that have been/need to be downloaded? Then each thread/async fn would lock the progress bar and update it every 1 MB or so.

**Comment by Mark-Simulacrum** [comment]
I've always been of the opinion that this shouldn't make downloads faster because we should be entirely limited by network bandwidth -- that said, it would be good to confirm this (e.g., with curl or something running in parallel).

**Comment by camelid** [comment]
Yeah, unfortunately my internet ends up being 4 MB/s download speed, so bisecting can take a while :/

I use `--preserve`, so that helps, but then I use up a ton of disk space.

**Comment by Mark-Simulacrum** [comment]
What we *can* do is parallelize compilation and downloading the next batch of artifacts. It would mean branch predicting that -- or downloading both -- but that seems not unreasonable. I guess if the compilation is fast enough it's not actually a win though.

**Comment by bjorn3** [comment]
Re branch predicting: I once read that it can sometimes be quicker to divide the range in three using two comparisons and branches rather than dividing the range in two. The former biases the first branch to not taken instead of the pathological case of taken and not taken being equally likely.
